### PR TITLE
allow mods to change the velocity of players

### PR DESCRIPTION
**Pull request by HybridDog**
it can optionally be set relatively to perform pushes

@sofar this can be used to make tnt push players

l added the set_player_velocity instead of making set_velocity work for players because of compatibility to #1489.

### CODE REVIEW

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -1073,6 +1073,20 @@ int ObjectRef::l_set_look_yaw(lua_State *L)
 	return 1;
 }
 
+// set_player_velocity(self, velocity, relative)
+int ObjectRef::l_set_player_velocity(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO* co = getplayersao(ref);
+	if (co == NULL) return 0;
```
Comment (sofar) [inline src/script/lua_api/l_object.cpp]:
```
if (!co)
        return 0;
```

File: src/content_sao.cpp [hunk 2]
```diff
@@ -1063,6 +1063,15 @@ void PlayerSAO::setPitch(float pitch)
 	((Server*)m_env->getGameDef())->SendMovePlayer(m_peer_id);
 }
 
+void PlayerSAO::setVelocity(v3f vel, bool relative)
+{
+	if(isAttached())
```
Comment (sofar) [inline src/content_sao.cpp]: space after `if`

**Comment by sofar** [comment]
it looks clean enough. Two style comments. Have not tested yet.

File: src/server.cpp [hunk 3]
```diff
@@ -1893,6 +1895,27 @@ void Server::SendMovePlayer(u16 peer_id)
...
+void Server::SendVelocityPlayer(u16 peer_id)
+{
+	DSTACK(FUNCTION_NAME);
+	Player *player = m_env->getPlayer(peer_id);
+	assert(player);
+
+	NetworkPacket pkt(TOCLIENT_VELOCITY_PLAYER, sizeof(v3f) + sizeof(bool), peer_id);
+	pkt << player->vel_change << player->vel_change_relative;
+
+	{
```
Comment (est31) [inline src/server.cpp]: the braces are not required.

**Comment by est31** [comment]
:-1: for reasons stated in #2960. There should be no way the server could influence the absolute velocity, and even relative effects should only be timed (probably smoothly approaching zero). Otherwise it just makes everything a total mess when the server lags.

**Comment by Foghrye4** [comment]
@est31 what if change velocity only on client side? Add two more variables to local player class - speed vector and time of action, send it from server and change LocalPlayer::applyControl so last two rows
accelerateHorizontal(speedH * physics_override_speed, incH * physics_override_speed);
accelerateVertical(speedV * physics_override_speed, incV * physics_override_speed);
will include our vector as well?

**Comment by paramat** [comment]
:-1: For the reasons est31 gives.

File: src/script/lua_api/l_object.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: `NULL` is valid because that's the return value of `getplayersao`.

**Pull request closed without merging**